### ISSUE TITLE
Which part of your code copies the encoder hidden when glancing?

### ISSUE THREAD

**Issue by zkx06111** [FLC777/GLAT#4]
Hi,
I read your paper and code with great interest. You said in the paper that during glancing, the sampled tokens are replaced with the embeddings from decoder, while the unsampled tokens use encoder output.
However, in the code, the unglanced tokens are still masked.
Am I missing something?
```python
       glat_info = None
        if glat and tgt_tokens is not None:
            with torch.no_grad():
                with torch_seed(rand_seed):
                    word_ins_out = self.decoder(
                        normalize=False,
                        prev_output_tokens=prev_output_tokens,
                        encoder_out=encoder_out,
                    )
                pred_tokens = word_ins_out.argmax(-1)
                same_num = ((pred_tokens == tgt_tokens) & nonpad_positions).sum(1)
                input_mask = torch.ones_like(nonpad_positions)
                bsz, seq_len = tgt_tokens.size()
                for li in range(bsz):
                    target_num = (((seq_lens[li] - same_num[li].sum()).float()) * glat['context_p']).long()
                    if target_num > 0:
                        input_mask[li].scatter_(dim=0, index=torch.randperm(seq_lens[li])[:target_num].cuda(), value=0)
                input_mask = input_mask.eq(1)
                input_mask = input_mask.masked_fill(~nonpad_positions,False)
                glat_prev_output_tokens = prev_output_tokens.masked_fill(~input_mask, 0) + tgt_tokens.masked_fill(input_mask, 0)
                # this line here
                
                glat_tgt_tokens = tgt_tokens.masked_fill(~input_mask, self.pad)

                prev_output_tokens, tgt_tokens = glat_prev_output_tokens, glat_tgt_tokens

                glat_info = {
                    "glat_accu": (same_num.sum() / seq_lens.sum()).item(),
                    "glat_context_p": glat['context_p'],
                }
```

**Comment by myaxxxxx** [comment]
I have several questions about GLAT. First, the author said the model would randomly replace mismatch tokens embedding with target token embeddings and decode again. However, 1) in your figure2, you replace h1, h3, and h5 with y1, y3, and y5, yet in the encoder-decoder framework, the length of the source and target sentence is always different. For example, if your source sentence length is five and target length is 7. Your mismatch token position is six; how to replace the hidden states?  2) I have read your code, it seems you modify mask operation in the decoder stage. After calculating the number of mismatch tokens called "target_num," you randomly mask target tokens according to "target_num" in the second decoding.

Your proposed model won the WMT21 DE-EN translation theme track, and it is a brilliant work for NAT. I think maybe random mask operation leads to great performance, and it masks several simple tokens and forces the model to train hard tokens.
